app/services/widgets/widget-source.ts

```diff
@@ -0,0 +1,82 @@
+import { ServiceHelper } from 'services/stateful-service';
+import { Inject } from '../../util/injector';
+import { Source, SourcesService } from 'services/sources';
+import { WidgetsService } from './widgets';
+import { IWidgetSource, WidgetType, IWidgetData } from './index';
+import { WidgetSettingsService } from 'services/widgets';
+import Utils from '../utils';
+import { mutation } from '../stateful-service';
+
+@ServiceHelper()
+export class WidgetSource implements IWidgetSource {
+
+  @Inject() private sourcesService: SourcesService;
+  @Inject() private widgetsService: WidgetsService;
+
+  readonly type: WidgetType;
+  readonly sourceId: string;
+  readonly previewSourceId: string;
+
+  private widgetSourceState: IWidgetSource;
+
+  constructor(sourceId: string) {
+    this.widgetSourceState = this.widgetsService.state.widgetSources[sourceId];
+    Utils.applyProxy(this, this.widgetSourceState);
+  }
+
+  getSource() {
+    return this.sourcesService.getSource(this.sourceId);
+  }
+
+  getSettingsService(): WidgetSettingsService<IWidgetData> {
+    return this.widgetsService.getWidgetSettingsService(this.type);
```

Comment:
where is type actually set here? or is it set by widgets that extend this?

Reply:
it is in `Utils.applyProxy(this, this.widgetSourceState);`
This function transforms `type`, `sourceId` and `previewSourceId` into getters for Vuex.store
So instances of `WidgetSource ` are always up to date with store. We use the same pattern for Sources and SceneItems

That works fine because of state item  is `IWidgetSource`  and `WidgetSource implements IWidgetSource` line